### DISCUSSION TITLE
[React] API：createContext

### DISCUSSION THREAD

**Discussion by kiinlam** [kiinlam/kiinlam.github.io#63] [04 React]
`createContext` 用于创建一个 `context` 以便组件能提供或获取设置的值。

```javascript
const SomeContext = createContext(defaultValue)
```

### 创建

在任意组件外调用 `createContext` 来创建一个上下文 `SomeContext`，可传入默认值 `defaultValue`。

```javascript
import { createContext } from 'react';

export const ThemeContext = createContext('light');
```

创建的 `context` 对象本身不包含任何信息。 它只表示其他组件读取或提供的特定 `context`。在组件上方使用 `SomeContext.Provider` 来指定 `context` 的值，并在被包裹的下方组件内调用 `useContext(SomeContext)` 来读取它。

### 指定 `context` 的值

通过 `value` prop 指定 `context` 的值。

```javascript
import { ThemeContext } from './Contexts.js';

function App() {
  const [theme, setTheme] = useState('light');
  // ...
  return (
    <ThemeContext.Provider value={theme}>
      <Page />
    </ThemeContext.Provider>
  );
}
```

### 读取 `context` 的值

通过 `useContext` hook 读取 `context` 的值。

```javascript
import { ThemeContext } from './Contexts.js';

function Button() {
  // 读取 context 的值
  const theme = useContext(ThemeContext);
  return <button className={theme} />;
}
```

参考：[React] Hook：useContext（#47）

**Comment by Aghaei-Dev** [comment]
You can use this sample :

```js
import React, {
  useState,
  useContext,
  useEffect,
  useCallback,
  createContext,
} from 'react'


export const AppContext = createContext()

export const AppProvider = ({ children }) => {
 
  const ctxVal = {
  }
  return <AppContext.Provider value={ctxVal}>{children}</AppContext.Provider>
}

export const useGlobalContext = () => useContext(AppContext)
```

just ready to use

from everywhere of your project and call the useGlobalContext and construct what You want!

**Reply by kiinlam** [reply]
If `ctxVal` will not change, that's great. Otherwise, you have to be careful.

Whenever `AppProvider` re-renders,`ctxVal` will be a different object, so React will also have to re-render all components deep in the tree that call `useGlobalContext`.

You could take a look at this chapter: [Optimizing re-renders when passing objects and functions](https://react.dev/reference/react/useContext#optimizing-re-renders-when-passing-objects-and-functions)
